lightclient/src/lib.rs

```diff
@@ -2,52 +2,259 @@
+/// it with [`LightClient::relay_chain()`] to communicate with a relay chain, and
+/// then call [`LightClient::parachain()`] to establish connections to parachains.
+#[derive(Clone)]
+pub struct LightClient {
+    client: SharedClient<DefaultPlatform>,
+    relay_chain_id: smoldot_light::ChainId,
+}
+
+impl LightClient {
+    /// Given a chain spec, establish a connection to a relay chain. Any subsequent calls to
+    /// [`LightClient::parachain()`] will set this as the relay chain.
+    ///
+    /// # Panics
+    ///
+    /// The panic behaviour depends on the feature flag being used:
+    ///
+    /// ## Native
+    ///
+    /// Panics when called outside of a `tokio` runtime context.
+    ///
+    /// ## Web
+    ///
+    /// If smoldot panics, then the promise created will be leaked. For more details, see
+    /// https://docs.rs/wasm-bindgen-futures/latest/wasm_bindgen_futures/fn.future_to_promise.html.
+    pub fn relay_chain<'a>(
+        chain_config: impl Into<ChainConfig<'a>>,
+    ) -> Result<(Self, LightClientRpc), LightClientError> {
+        let mut client = smoldot_light::Client::new(platform::build_platform());
+        let chain_config = chain_config.into();
+        let chain_spec = chain_config.as_chain_spec();
+
+        let config = smoldot_light::AddChainConfig {
+            specification: chain_spec,
+            json_rpc: smoldot_light::AddChainConfigJsonRpc::Enabled {
+                max_pending_requests: u32::MAX.try_into().unwrap(),
+                max_subscriptions: u32::MAX,
+            },
+            database_content: "",
+            potential_relay_chains: std::iter::empty(),
+            user_data: (),
+        };
+
+        let added_chain = client
+            .add_chain(config)
+            .map_err(|err| LightClientError::AddChainError(err.to_string()))?;
+
+        let relay_chain_id = added_chain.chain_id;
+        let rpc_responses = added_chain
+            .json_rpc_responses
+            .expect("Light client RPC configured; qed");
+        let shared_client: SharedClient<_> = client.into();
+
+        let light_client_rpc =
+            LightClientRpc::new_raw(shared_client.clone(), relay_chain_id, rpc_responses);
+        let light_client = Self {
+            client: shared_client,
+            relay_chain_id,
+        };
+
+        Ok((light_client, light_client_rpc))
+    }
+
+    /// Given a chain spec, establish a connection to a parachain.
+    ///
+    /// # Panics
+    ///
+    /// The panic behaviour depends on the feature flag being used:
+    ///
+    /// ## Native
+    ///
+    /// Panics when called outside of a `tokio` runtime context.
+    ///
+    /// ## Web
+    ///
+    /// If smoldot panics, then the promise created will be leaked. For more details, see
+    /// https://docs.rs/wasm-bindgen-futures/latest/wasm_bindgen_futures/fn.future_to_promise.html.
+    pub fn parachain<'a>(
+        &self,
+        chain_config: impl Into<ChainConfig<'a>>,
```

Comment:
This interface with the `relay_chain` and `parachain` functions looks really clean.